barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_group.cpp

```diff
@@ -0,0 +1,1322 @@
+#include "../field/field.hpp"
+#include "barretenberg/crypto/pedersen_commitment/pedersen.hpp"
+#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
+
+#include "../../hash/pedersen/pedersen.hpp"
+#include "../../hash/pedersen/pedersen_gates.hpp"
+
+#include "./cycle_group.hpp"
+#include "barretenberg/proof_system/plookup_tables/types.hpp"
+#include "barretenberg/stdlib/primitives/plookup/plookup.hpp"
+namespace proof_system::plonk::stdlib {
+
+template <typename Composer>
+cycle_group<Composer>::cycle_group(Composer* _context)
+    : context(_context)
+    , x(0)
+    , y(0)
+    , _is_infinity(true)
+    , _is_constant(true)
+{}
+
+/**
+ * @brief Construct a new cycle group<Composer>::cycle group object
+ *
+ * @param _x
+ * @param _y
+ * @param is_infinity
+ */
+template <typename Composer>
+cycle_group<Composer>::cycle_group(field_t _x, field_t _y, bool_t is_infinity)
+    : context(_x.get_context() == nullptr
+                  ? _y.get_context() == nullptr
+                        ? is_infinity.get_context() == nullptr ? nullptr : is_infinity.get_context()
+                        : _y.get_context()
+                  : _x.get_context())
+    , x(_x.normalize())
+    , y(_y.normalize())
+    , _is_infinity(is_infinity)
+    , _is_constant(_x.is_constant() && _y.is_constant() && is_infinity.is_constant())
+{}
+
+/**
+ * @brief Construct a new cycle group<Composer>::cycle group object
+ *
+ * @details is_infinity is a circuit constant. We EXPLICITLY require that whether this point is infinity/not infinity is
+ * known at circuit-construction time *and* we know this point is on the curve. These checks are not constrained.
+ * Use from_witness if these conditions are not met.
+ * Examples of when conditions are met: point is a derived from a point that is on the curve + not at infinity.
+ * e.g. output of a doubling operation
+ * @tparam Composer
+ * @param _x
+ * @param _y
+ * @param is_infinity
+ */
+template <typename Composer>
+cycle_group<Composer>::cycle_group(const FF& _x, const FF& _y, bool is_infinity)
+    : context(nullptr)
+    , x(_x)
+    , y(_y)
+    , _is_infinity(is_infinity)
+    , _is_constant(true)
+{
+    ASSERT(get_value().on_curve());
+}
+
+/**
+ * @brief Construct a cycle_group object out of an AffineElement object
+ *
+ * @note This produces a circuit-constant object i.e. known at compile-time, no constraints.
+ *       If `_in` is not fixed for a given circuit, use `from_witness` instead
+ *
+ * @tparam Composer
+ * @param _in
+ */
+template <typename Composer>
+cycle_group<Composer>::cycle_group(const AffineElement& _in)
+    : context(nullptr)
+    , x(_in.x)
+    , y(_in.y)
+    , _is_infinity(_in.is_point_at_infinity())
+    , _is_constant(true)
+{}
```

Comment:
To be clear, for SW curves, the point at infinity cannot be encoded as an affine element -- This is different to (0,0) which is a point not on the curve but is usually denoted as being the point at infinity in software, some libraries use it for performance assuming new points are zeroed out.

Can you explain whether its possible for the following cases:

- (0,0) is_inf = true
- (x,y) is_inf = true [x,y] is on the curve
- (x,y) is_inf = false [x,y] is on the curve

Reply:
All three cases are possible.

if `is_inf = true`, the code will treat the point as a point at infinity regardless of the contents of the `x, y` coordinates